### ISSUE TITLE
redundant_slicing false positive & bad suggestion

### ISSUE THREAD

**Issue by CPerezz** [rust-lang/rust-clippy#7113]
Lint name: redundant_slicing

I tried this code:

```rust
/// Deserialise a slice of bytes into a Public Parameter struct performing
    /// security and consistency checks for each point that the bytes
    /// contain.
    ///
    /// # Note
    /// This function can be really slow if the `PublicParameters` have a
    /// certain degree. If the bytes come from a trusted source such as a
    /// local file, we recommend to use [`PublicParameters::
    /// from_slice_unchecked`] and [`PublicParameters::to_raw_var_bytes`].
    pub fn from_slice(bytes: &[u8]) -> Result<PublicParameters, Error> {
        if bytes.len() <= OpeningKey::SIZE {
            return Err(Error::NotEnoughBytes);
        }
        let mut buf = &bytes[..];
        let opening_key = OpeningKey::from_reader(&mut buf)?;
        let commit_key = CommitKey::from_slice(&buf)?;

        let pp = PublicParameters {
            opening_key,
            commit_key,
        };

        Ok(pp)
    }
```
I don't think it is really important the fact of what the code is, but indeed what it tries to do. 

We try to call `from_reader`. Which needs to get a `&mut [u8]` and reads from it X bytes so that you can use it later and continue reading from it.
To set up the reader, I need to create a mutable variable that references the borrowed `bytes` and I'm forced to slice the whole range.

Clippy complains and suggests the following:
```
         let mut buf = &bytes[..];
                       ^^^^^^^^^^ help: use the original slice instead: `bytes`

```
This does not work, since if I remove the reference from `bytes` the error about size not known at compile-time appears as is normal.

### Meta

- `cargo clippy -V`: clippy 0.1.52 (07e0e2e 2021-03-24)
- `rustc -Vv`:

  ```
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
binary: rustc
commit-hash: 07e0e2ec268c140e607e1ac7f49f145612d0f597
commit-date: 2021-03-24
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

**Comment by camsteffen** [comment]
I think you did not remove the `[..]`?

**Comment by CPerezz** [comment]
You're absolutely right @camsteffen . On the way I read the message, I interpreted that it was suggesting to remove the `&` instead of the `[..]`.

`[..]` is completely redundant. Apologies for the noise and thanks!

My only suggestion would be to keep the `&` in the err message. Otherways can be missinterpreted although I should have catched it..

**Comment by camsteffen** [comment]
No problem!

The `&` and the `[..]` need to be removed together since `[..]` creates a `[T]` which needs to be referenced. On the other hand, keeping the `&` and removing the `[..]` would leave you with a double reference `&&[T]`.
